x/wasm/keeper/keeper.go

```diff
@@ -565,7 +567,30 @@ func (k Keeper) callMigrateEntrypoint(
 	prefixStoreKey := types.GetContractStorePrefix(contractAddress)
 	vmStore := types.NewStoreAdapter(prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(sdkCtx)), prefixStoreKey))
 	gasLeft := k.runtimeGasForContract(sdkCtx)
-	res, gasUsed, err := k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+
+	migrateInfo := wasmvmtypes.MigrateInfo{
+		Sender:            senderAddress.String(),
+		OldMigrateVersion: OldMigrateVersion,
+	}
+	var res *wasmvmtypes.ContractResult
+	var gasUsed uint64
+	var err error
+	if OldMigrateVersion == nil {
+		res, gasUsed, err = k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+		if err != nil {
+			if strings.Contains(err.Error(), "The called function args arity does not match") {
+				res, gasUsed, err = k.wasmVM.MigrateWithInfo(newChecksum, env, msg, migrateInfo, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+			}
+		}
+	} else {
+		res, gasUsed, err = k.wasmVM.MigrateWithInfo(newChecksum, env, msg, migrateInfo, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+		if err != nil {
+			if strings.Contains(err.Error(), "The called function args arity does not match") {
+				res, gasUsed, err = k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+			}
+		}
+	}
```

Comment:
As mentioned in the call today, I suggest here to always call `k.wasmVM.MigrateWithInfo(` with the additional information and then do the error handling in the Rust land (probably libwasmvm)

Reply:
Yes, that's a good solution. I've added the migrate function arity check in cosmwasm-vm, so it is handled purely on the Rust side now.